src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java

```diff
@@ -247,31 +278,91 @@ private void submitTopicReadOperation() {}
 
   private void submitTopicDeleteOperation() {}
 
+  private int submitBatchRecordRead(final List<O> ops, final int from, final int to) {
+    var submitCount = 0;
+    if (from < to && concurrencyThrottle.tryAcquire()) {
+      try {
+        val consumerConfig = new Properties();
+        val anyOp = ops.get(from);
+        val nodeAddr = anyOp.nodeAddr();
+        val amountOfRecords = to - from;
+        consumerConfig.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, nodeAddr);
+        consumerConfig.put(
+            ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+            "org.apache.kafka.common.serialization.StringDeserializer");
+        consumerConfig.put(
+            ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+            "org.apache.kafka.common.serialization.ByteArrayDeserializer");
+        consumerConfig.put(ConsumerConfig.SEND_BUFFER_CONFIG, this.sndBuf);
+        consumerConfig.put(ConsumerConfig.RECEIVE_BUFFER_CONFIG, this.rcvBuf);
+        consumerConfig.put(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, String.valueOf(amountOfRecords));
+        consumerConfig.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
+        final KafkaConsumer<String, byte[]> consumer = new KafkaConsumer<>(consumerConfig);
+        consumer.subscribe(Arrays.asList(anyOp.dstPath())); // Could ops have different dstPaths?
+        List<Integer> sizesOfReadRecords = new ArrayList<>(amountOfRecords);
+        for (var i = from; i < to; i++) {
+          O recordOp = ops.get(i);
+          recordOp.startRequest();
+        }
+        val readRecords = consumer.poll(Duration.ZERO);
+        for (ConsumerRecord<String, byte[]> record : readRecords) {
+          if (record != null) {
+            sizesOfReadRecords.add(record.value().length);
+          }
+        }
+        if (sizesOfReadRecords.size() >= amountOfRecords) {
+          completeRecordReadOperations(ops, from, to, SUCC, sizesOfReadRecords);
+          submitCount = amountOfRecords;
+          consumer.commitSync();
```

Comment:
why do we need this?

Reply:
I think that if we failed to read at least one operation from the list we need to roll back the offset. Am I wrong?

Reply:
but then we could read some records twice, right?

Reply:
Hm. Ok then we can commitSync after each correctly read record

Reply:
Nope. We can't use it like this. Because we need to commitSync after poll method only.